### PR TITLE
Add envoy-json-logging-custom-fields-design doc

### PR DESCRIPTION
**Pull request by mike1808**
Co-authored-by: Leah Hanson <lhanson@pivotal.io>

In continuation of #3032 we're making a design doc to discuss possible solutions for our problem.

### CODE REVIEW

File: design/envoy-json-logging-custom-fields-design.md
Comment (jpeach) [inline design/envoy-json-logging-custom-fields-design.md:143]: Map syntax seems clean, but we would need to introduce it in a new config field, then deprecate the old field after a number of releases. Would still need to define what happens to the built-in field mappings.

File: design/envoy-json-logging-custom-fields-design.md
Comment (jpeach) [inline design/envoy-json-logging-custom-fields-design.md:139]: The built-in mappings still need to be supported. I would not want to force people using the existing field mappings to manually remake their configuration.

File: design/envoy-json-logging-custom-fields-design.md
Comment (jpeach) [inline design/envoy-json-logging-custom-fields-design.md:61]: See https://github.com/projectcontour/contour/pull/3023 for config file validation.

**Review by jpeach — APPROVED** [review]
This looks good to me. I'll give other maintainers a chance to review before I merge though.

**Review by youngnick — APPROVED** [review]
This proposal LGTM, I think it's a good balance between adding extra functionality and backward compatibility. Thanks for writing it up @mike1808 and @astrieanna!

File: design/envoy-json-logging-custom-fields-design.md
Comment (astrieanna) [reply design/envoy-json-logging-custom-fields-design.md:139]: We agree that built-in mappings need to be supported. The question here is whether fields that could be added to those mappings should be.
* `method` will still work.
* should `response_duration` work?

If `response_duration` works, then the interface is consistent, but the implication is that the translation table will be used forever (despite users having a way to write `response_duration` without the translation table).
If `response_duration=%RESPONSE_DURATION%` is the only option for users, then there is inconsistency between the original set that are in the translations and the new ones that require more spelling out.

File: design/envoy-json-logging-custom-fields-design.md
Comment (astrieanna) [reply design/envoy-json-logging-custom-fields-design.md:143]: Yea, since the map syntax would require deprecating the prior configuration field anyway, the new field would not have to support the built-in field mappings. In our example in this doc, we assumed that empty values would imply that the key should be looked up in the built-in field mappings, but I'm not sure whether that's desirable or not.

## In favor of a built-in translation table:
* Users of Contour only have to read Contour's docs (vs. also referencing Envoy's config syntax).
* Users get reasonable, standardized field names because that's what's setup by the field mappings.

## In favor of writing all key-value pairs out:
* Users of Contour can read the Envoy docs to decide what config to write. They don't need to know Contour's specific translation of special keys.

Comment (youngnick) [reply design/envoy-json-logging-custom-fields-design.md:143]: I like the ease of use of having the translation table, personally. I think that having the two types seems like an okay tradeoff for that ease of use (for now at least).

File: design/envoy-json-logging-custom-fields-design.md
Comment (youngnick) [reply design/envoy-json-logging-custom-fields-design.md:139]: The original intent was that we add simple translations as required to the translation table; we figured that people could just request anything they needed added. With that in mind, I'm okay with adding whatever simple translations to the translation table (like `response_duration`, where it's a straight mapping between field name and Envoy template string), as part of this implementation.

Obviously, this will be frozen at this point in time, but perhaps a comment saying that adding things to the standard translations is fine would make it clearer that we're open to adding more?

File: design/envoy-json-logging-custom-fields-design.md
Comment (stevesloka) [inline design/envoy-json-logging-custom-fields-design.md:57]: Wonder if this should raise a validation error instead of overwriting?

**Review by skriss — APPROVED** [review]
LGTM!

File: design/envoy-json-logging-custom-fields-design.md
Comment (jpeach) [reply design/envoy-json-logging-custom-fields-design.md:57]: A validation error could be OK if a field name is repeated in a single `json-fields` array. My original intention with suggesting overwriting was that operators should be able to use this mechanism to change the definition of builtin log fields (ie. overwriting the builtin field mapping).

**Comment by jpeach** [comment]
Last call for comments.

File: design/envoy-json-logging-custom-fields-design.md
Comment (youngnick) [inline design/envoy-json-logging-custom-fields-design.md:3]: If there are no further comments, @mike1808, could you change this to 'Accepted' before we merge?

Comment (mike1808) [reply design/envoy-json-logging-custom-fields-design.md]: Done @youngnick

**Comment by youngnick** [comment]
Okay, I'm calling this one done. Thanks @mike1808 and @astrieanna!

**Pull request merged**
